documentation/src/main/asciidoc/backend-lucene.asciidoc

```diff
@@ -14,14 +12,184 @@ but the defaults might not suit everyone.
+See `org.apache.lucene.store.SimpleFSDirectory`.
+* `mmap`: uses `mmap` for reading, and `FSDirectory.FSIndexOutput` for writing.
+See `org.apache.lucene.store.MMapDirectory`.
+* `nio`: uses ``java.nio.channels.FileChannel``'s positional read for concurrent reading,
+and `FSDirectory.FSIndexOutput` for writing.
+See `org.apache.lucene.store.NIOFSDirectory`.
+
+[NOTE]
+====
+Make sure to refer to Javadocs of these `Directory`
+implementations before changing this setting.
+Implementations offering better performance
+also bring issues of their own.
+====
+
+===== Other configuration options
+
+The `local-filesystem` directory also allows configuring a
+<<backend-lucene-configuration-directory-locking-strategy,locking strategy>>.
+
+[[backend-lucene-configuration-directory-local-heap]]
+==== Local heap storage
+
+The `local-heap` directory type will store indexes in the local JVM's heap.
+
+As a result, indexes contained in a `local-heap` directory are *lost when the JVM shuts down*.
+
+This directory type is only provided for use in *testing configurations*
+with *small indexes* and *low concurrency*,
+where it could slightly improve performance.
+In setups requiring larger indexes and/or high concurrency,
+a <<backend-lucene-configuration-directory-local-filesystem,filesystem-based directory>>
+will achieve better performance.
+
+The `local-heap` directory does not offer any specific option
+beyond the <<backend-lucene-configuration-directory-locking-strategy,locking strategy>>.
+
+[[backend-lucene-configuration-directory-locking-strategy]]
+==== Locking strategy
+// Search 5 anchors backward compatibility
+[[search-configuration-directory-lockfactories]]
+
+In order to write to an index, Lucene needs to acquire a lock to ensure no other application instance
+writes to the same index concurrently.
+Each directory type comes with a default locking strategy that should work well enough in most situations.
+
+For those (very) rare situations where a different locking strategy is needed,
+Hibernate Search exposes a configuration property at the backend level:
+
+[source]
+----
+hibernate.search.backends.<backend-name>.directory.locking.strategy = /path/to/my/root
+----
 
-// TODO HSEARCH-3440 document this
+The following strategies are available:
+
+* `simple-filesystem`:
+Locks the index by creating a marker file and checking it before write operations.
+This implementation is very simple and based Java's File API.
+If for some reason an application ends abruptly,
+the marker file will stay on the filesystem and will need to be removed manually.
++
+See `org.apache.lucene.store.SimpleFSLockFactory`.
+* `native-filesystem`:
+Similarly to `simple-filesystem`, locks the index by creating a marker file,
+but using native OS file locks instead of Java's File API,
+so that locks will be cleaned up if the application ends abruptly.
++
+This implementation has known problems with NFS: it should be avoided on network shares.
```

Comment:
Didn't we say that network file systems are to be avoided in any cases?

Reply:
Right, that was clumsy. NFS can be used, but not to share data between multiple instances of Search. It could still be used to have Search run on one machine and the indexes be stored on another (though I doubt performance would be great).

I changed wording a bit to make this clearer.